lua/starfall/libs_cl/midi.lua

```diff
@@ -0,0 +1,103 @@
+if not SF.Require("midi") then return function() end end
+
+local checkluatype = SF.CheckLuaType
+
+--- Midi Library
+-- Polls midi event information from midi devices.
+-- @name midi
+-- @class library
+-- @libtbl midi_library
+SF.RegisterLibrary("midi")
+
+return function(instance)
+-- Only usable by the owner of the starfall chip
+if LocalPlayer() ~= instance.player then return end
+
+local midi_library = instance.Libraries.midi
+
+-- Close all ports when SF chip is deleted
+-- Ensures that the midi port can still be used in other applications after the SF chip is deleted
+instance:AddHook("deinitialize", function()
+	midi_library.closeAllPorts()
+end)
+
+--- Event hook for midi devices.  
+-- Everytime a midi device outputs a signal, the callback function on the hook is called.
+-- Read up on the MIDI protocol to make better sense of everything -> https://ccrma.stanford.edu/~craig/articles/linuxmidi/misc/essenmidi.html
+-- @name midi
+-- @class hook
+-- @client
+-- @libtbl midi_library
+-- @param number time the exact systime which the event occured
+-- @param number command the command code of the event.  First 4 bits are the command code and last 4 are the channel
+-- @param number param1 Each command has their own set of parameters, see above
+-- @param number param2 Each command has their own set of parameters, see above
+-- Commands and their parameters:
+-- 0x80 NOTE_OFF              : param1 = key;                         param2 = velocity
+-- 0x90 NOTE_ON               : param1 = key;                         param2 = velocity
+-- 0xA0 AFTERTOUCH            : param1 = key;                         param2 = touch
+-- 0xB0 CONTINUOUS_CONTROLLER : param1 = button_number;               param2 = button_value
+-- 0xC0 PATCH_CHANGE          : param1 = patch number;
+-- 0xD0 CHANNEL_PRESSURE      : param1 = pressure;
+-- 0xE0 PITCH_BEND            : param1 = lsb(least signifigant bit);  param2 = msb(most signifigant bit)
+SF.hookAdd("MIDI", "midi")
+
+--- Opens the midi port to make it available to grab events from.  This must be called before the hook.
+-- @param number port the midi port to open. Passing nothing defaults to 0.
+-- @return string the name of the midi device opened at the given port.
+function midi_library.openPort(port)
+	checkluatype(port, TYPE_NUMBER)
+	if midi_library.isPortOpen(port) then return end
```

Comment:
Think this is the only thing that needs to be changed now, like I said earlier we shouldn't fail silently, either keep the SF.Throw that you put earlier or preferably remove this check

Also docs are generated in a number of ways:
1. Host a local server with your fork of SF mounted and open the sfhelper in game
2. Generate docs with github actions (By default only generates when your main branch changes, so you'd need to change that in your fork.)
3. Generate them manually https://github.com/thegrb93/StarfallEx/wiki/Building-docs-manually (Just do 1 at this point)

Reply:
Now i see what you mean.  Ill fix that

Reply:
Earlier I thought you meant to add the SF throw for the closing of the port.  I had it silently fail because when adding a function to close everything it would throw within the binary.  So when the SF chip was deleted it would just throw an error, because i have it closing everything in the de initialize hook, to make sure that no ports are being held.